src/remote/workspace.py

```diff
@@ -0,0 +1,162 @@
+                       commands with simple will be executed from user's remote home directory
+        :param dry_run: log the command to be executed but don't run it
+        """
+        if isinstance(command, list):
+            command = self._prepare_command(command)
+        if not simple:
+            command = self._generate_command(command)
+
+        subprocess_command = ["ssh", "-tKq", self.remote.host, command]
+        logger.info("Executing:\n%s %s %s <<EOS\n%sEOS", *subprocess_command)
+        if dry_run:
+            return
+
+        start = time.time()
+        result = subprocess.run(subprocess_command, stdout=sys.stdout, stderr=sys.stderr, stdin=sys.stdin)
+        runtime = time.time() - start
+        logger.info("Execution done in %.2f seconds", runtime)
+        # ssh exits with the exit status of the remote command or with 255 if an error occurred
+        if result.returncode == 255:
+            raise RemoteConnectionError(f"Failed to connect to {self.remote.host}")
+        elif result.returncode != 0:
+            raise RemoteExecutionError(
+                f'Failed to execute "{command}" on host {self.remote.host} ({result.returncode})'
+            )
+
+    def push(self, info=False, verbose=False, dry_run=False, mirror=False):
+        """Push local workspace files to remote directory
+
+        :param info: use info logging when running rsync
+        :param verbose: use verbose logging when running rsync
+        :param dry_run: use dry_run parameter when running rsync
+        :param mirror: mirror local files remotely. It will remove ALL the remote files in the directory
+                       that weren't synced from local workspace
+        """
+        src = f"{self.local_root}/"
+        dst = f"{self.remote.host}:{self.remote.directory}"
+        ignores = self.ignores.compile_push_ignores()
+        rsync(src, dst, info=info, verbose=verbose, dry_run=dry_run, mirror=mirror, excludes=ignores)
+
+    def pull(self, info=False, verbose=False, dry_run=False, subpath=None):
+        """Pull remote files to local workspace
+
+        :param info: use info logging when running rsync
+        :param verbose: use verbose logging when running rsync
+        :param dry_run: use dry_run parameter when running rsync
+        :param subpath: a specific path to bring in. If provided, subpath will be synced
+                        even if it is ignored by workspace rules
+        """
+        if subpath is not None:
+            src = f"{self.remote.host}:{self.remote.directory}/{subpath}"
+            dst = str(self.local_root / subpath)
+            rsync(src, dst, info=info, verbose=verbose, dry_run=dry_run)
+            return
+
+        src = f"{self.remote.host}:{self.remote.directory}/"
+        dst = str(self.local_root)
+        ignores = self.ignores.compile_pull_ignores()
+        # TODO: figure out why it was used in shell version
+        # extra_args = ["--include='*/'", "--exclude='*'"]
```

Comment:
yes this can be dropped. I was playing around with different rsync options to get it to do what I needed it to do.